File: solved/problem19.py

def increment(year, month, day):
	day += calc_num_days_in_month(year, month)

	month %= 12
	month += 1

	if month == 1:
		year += 1
	
	return (year, month, day)

def calc_num_days_in_month(year, month):
	if month == 1: return 31
	elif month == 2:
		if year % 4 == 0 and (year % 100 != 0 or year % 400 == 0):
			return 29
		else:
			return 28
	elif month == 3: return 31
	elif month == 4: return 30
	elif month == 5: return 31
	elif month == 6: return 30
	elif month == 7: return 31
	elif month == 8: return 31
	elif month == 9: return 30
	elif month == 10: return 31
	elif month == 11: return 30
	elif month == 12: return 31

num_sundays = 0
year = 1901
month = 1
day = 2 # Tuesday
while year < 2000:
	if day % 7 == 0:
		num_sundays += 1

	(year, month, day) = increment(year, month, day)

print num_sundays + 1 #off-by-one somewhere

File: src/solved/problem19.hs

increment :: (Integer, Integer, Integer) -> (Integer, Integer, Integer)
increment (year, month, day) = (year', month', day')
	where
		day' :: Integer
		day' = day + (num_days_in_month year month)

		month' :: Integer
		month' = (month `mod` 12) + 1

		year' :: Integer
		year' = if month' == 1
			then year + 1
			else year

is_leap_year :: Integer -> Bool
is_leap_year year =
	((year `mod` 4) == 0) &&
	(((year `mod` 100) /= 0) || ((year `mod` 400) == 0))

num_days_in_month :: Integer -> Integer -> Integer
num_days_in_month year month =
	case month of
		1 -> 31
		2 -> if is_leap_year year
			then 29
			else 28
		3 -> 31
		4 -> 30
		5 -> 31
		6 -> 30
		7 -> 31
		8 -> 31
		9 -> 30
		10 -> 31
		11 -> 30
		12 -> 31

is_sunday :: Integer -> Bool
is_sunday day = (day `mod` 7) == 0

num_sundays :: Int
num_sundays
	= (+) 1
	. sum
	. map (\(y,m,d) -> if is_sunday d then 1 else 0)
	. takeWhile (\(y,m,d) -> y /= 2000)
	. iterate increment
	$ (1901, 1, 2)

main :: IO ()
main = do
	putStrLn . show $ num_sundays

